Zend/zend_compile.c

```diff
@@ -4712,6 +4712,125 @@ static void zend_compile_ns_call(znode *result, znode *name_node, zend_ast *args
 }
 /* }}} */
 
+static zend_op *zend_compile_rope_add(znode *result, uint32_t num, znode *elem_node);
+static zend_op *zend_compile_rope_add_ex(zend_op *opline, znode *result, uint32_t num, znode *elem_node);
+static void zend_compile_rope_finalize(znode *result, uint32_t j, zend_op *init_opline, zend_op *opline);
+
+static zend_result zend_compile_func_sprintf(znode *result, zend_ast_list *args) /* {{{ */
+{
+	znode format_string;
+
+	if (args->children < 1) {
+		return FAILURE;
+	}
+
+	zend_compile_expr(&format_string, args->child[0]);
+	if (!(format_string.op_type == IS_CONST && Z_TYPE(format_string.u.constant) == IS_STRING)) {
+		zend_do_free(&format_string);
+		return FAILURE;
+	}
+	if (!(Z_STRLEN(format_string.u.constant) < 32)) {
+		zval_ptr_dtor_str(&format_string.u.constant);
+		return FAILURE;
+	}
+
+	char *p;
+	char *end;
+	uint32_t string_placeholder_count;
+	
+	string_placeholder_count = 0;
+	p = Z_STRVAL(format_string.u.constant);
+	end = p + Z_STRLEN(format_string.u.constant);
+	while (p < end) {
+		if (*p != '%') {
+			p++;
+			continue;
+		}
+		char *q = p + 1;
+		if (q == end) {
+			zval_ptr_dtor_str(&format_string.u.constant);
+			return FAILURE;
+		}
+		if (*q == '%') {
+			p++;
+			continue;
+		}
+
+		if (*q != 's') {
+			zval_ptr_dtor_str(&format_string.u.constant);
+			return FAILURE;
+		}
+
+		string_placeholder_count++;
+
+		p = q;
+		p++;
+	}
+
+	if (string_placeholder_count != (args->children - 1)) {
+		zval_ptr_dtor_str(&format_string.u.constant);
+		return FAILURE;
+	}
+
+	uint32_t rope_elements = 0;
+	uint32_t rope_init_lineno = -1;
+	zend_op *opline;
+
+	string_placeholder_count = 0;
+	p = Z_STRVAL(format_string.u.constant);
+	end = p + Z_STRLEN(format_string.u.constant);
+	char *offset = p;
+	while (p < end) {
+		znode elem_node;
+		if (*p != '%') {
+			p++;
+			continue;
+		}
+		char *q = p + 1;
+		ZEND_ASSERT(q < end);
+		if (*q == '%') {
+			p++;
+			continue;
+		}
+		ZEND_ASSERT(*q == 's');
+
+		zend_compile_expr(&elem_node, args->child[1 + string_placeholder_count]);
+		if (rope_elements == 0) {
+			rope_init_lineno = get_next_op_number();
+		}
+		if (p != offset) {
+			znode const_node;
+			const_node.op_type = IS_CONST;
+			ZVAL_STRINGL(&const_node.u.constant, offset, p - offset);
+			opline = zend_compile_rope_add(result, rope_elements++, &const_node);
```

Comment:
You should probably compile all the list items before adding them to the rope. For `sprintf`, all arguments are evaluated before calling the function. https://3v4l.org/cojjT

Reply:
Good catch, fixed. Is there a smart way to avoid the dynamic memory allocation to store the resulting pointers?

Reply:
You could try `do_alloca`, which uses `alloca` when possible, and `emalloc` otherwise.

Reply:
That one also requires manual freeing in case a dynamic allocation is used and doesn't provide any “multiplication overflow safety”. I was hoping for some way to work with the existing information.

I'll leave it as is, thank you.